lxd/main_init.go

```diff
@@ -494,8 +510,148 @@ they otherwise would.
 			return err
 		}
 	}
+	return nil
+}
+
+// Run the logic for preseed mode
+func (cmd *CmdInit) runPreseed(client lxd.ContainerServer) error {
+	data := &cmdInitData{}
+
+	err := cmd.Context.InputYAML(data)
+	if err != nil {
+		return fmt.Errorf("Invalid preseed YAML content")
+	}
+
+	return cmd.run(client, data)
+}
+
+// Apply the configuration specified in the given init data.
+func (cmd *CmdInit) run(client lxd.ContainerServer, data *cmdInitData) error {
```

Comment:
Hm if we can't revert completely, I'm not sure to see the value of partial rollback vs clearly stating "You asked me to do X, Y, and Z, I got to X successfully but then Y failed, so I'm giving up".

I assume one change that can't be easily reverted is storage pool creation, which I believe involves some creation of external assets such as zfs/btrs/lvm volumes. What would reverting look like in that case? Delete the state from our db but keep the external assets?

Also, the current behavior of lxd init is not transactional as well (it can fail midway).

Happy to get more input on this, if there's a reasonable and not surprising way to get transactional-ish behavior.